src/adera/SysExhaustPlume.cpp

```diff
@@ -0,0 +1,82 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2020 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "SysExhaustPlume.h"
+#include "osp/Active/ActiveScene.h"
+#include "adera/Machines/Rocket.h"
+#include "osp/Resource/AssetImporter.h"
+#include "osp/Active/SysDebugRender.h"
+
+#include <Magnum/Trade/MeshData.h>
+#include <Magnum/MeshTools/Compile.h>
+#include <Magnum/GL/Mesh.h>
+#include <Magnum/GL/Texture.h>
+#include <Magnum/Trade/ImageData.h>
+#include <Magnum/ImageView.h>
+#include <Magnum/GL/Sampler.h>
+#include <Magnum/GL/TextureFormat.h>
+
+using namespace osp::active;
+
+const std::string SysExhaustPlume::smc_name = "ExhaustPlume";
+
+SysExhaustPlume::SysExhaustPlume(ActiveScene& scene)
+    : m_scene(scene), m_time(0.0f),
+    m_updatePlume(scene.get_update_order(), "exhaust_plume", "mach_rocket", "",
+        std::bind(&SysExhaustPlume::update_plumes, this))
+{}
+
+void SysExhaustPlume::update_plumes()
+{
+    m_time += m_scene.get_time_delta_fixed();
+
+    using adera::active::machines::MachineRocket;
+    using ShaderInstance_t = adera::shader::PlumeShader::ACompPlumeShaderInstance;
+
+    auto plumeView = m_scene.get_registry()
+        .view<ACompExhaustPlume, ShaderInstance_t, CompVisibleDebug>();
+
+    for (ActiveEnt plumeEnt : plumeView)
+    {
+        auto& plume = plumeView.get<ACompExhaustPlume>(plumeEnt);
+        auto& plumeShader = plumeView.get<ShaderInstance_t>(plumeEnt);
+        CompVisibleDebug& visibility = plumeView.get<CompVisibleDebug>(plumeEnt);
+
+        auto& machine = m_scene.reg_get<MachineRocket>(plume.m_parentMachineRocket);
```

Comment:
I'm going to assume most of this is temporary, but rocket and plumes both depend on each other. Rockets have to initialize the plumes, and plumes are reading MachineRockets specifically.

A possible solution would be to add a new component for plume engine power, in the same entity as MachineRocket. MachineRocket can then write its throttle value into it. This allows other systems aside that aren't rockets to use plumes.

Reply:
Systems like what? I suppose jet engines might be a different type of system? I figured that more general effects would have their own system, which is why this system and shader are designed so specifically for rocket exhaust plumes

Reply:
I consider it generally cleaner to separate parts of the code when possible. Machines are designed to be replaceable by having the same wire connection names, almost like duck typing. It's likely that MachineRocket won't become a ultra-generalized to fit every rocket model possible. Someone might want their own MachineRealisticRocket.

Reply:
Hm ok. For now, a good middle ground would be to just have `ACompExhaustPlume` store the throttle value, and then make the MachineRocket or whatever other propulsion system write the throttle value, like you mentioned. That makes sense to me since SysExhaustPlume is purely graphics-oriented, so making this change would make it manage only the rendering/shader and be completely agnostic to how it gets its information. Does that sound ok?

Reply:
That's actually a much cleaner and more decoupled way, and I'm not sure how I missed it originally

Reply:
If MachineRocket writes directly into its plumes, then this adds the issue of having to loop through them. I suggested a separate component to prevent this from happening. Also, engine power can also be made into a wire output. I might make some changes to the wire system that will make this easier to do, so it might be okay to just leave this as-is.

Reply:
Alright, if you're ok with it as-is I'll leave it and we can discuss/experiment and change it when we have a plan in place, or if any clear solutions present themselves when things progress.